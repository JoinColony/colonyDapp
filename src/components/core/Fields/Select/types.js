/* @flow */

export type Appearance = {
  alignOptions?: 'left' | 'center' | 'right',
  theme?: 'default' | 'alt',
};
