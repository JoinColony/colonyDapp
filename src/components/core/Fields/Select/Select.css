.baseSelect {
  padding: 5px 10px;
  width: auto;
  width: 100%;
  position: relative;
  border: 0;
  font-size: var(--size-normal);
  font-style: normal;
  color: var(--text);
  transition: none;
  cursor: pointer;
  outline-color: var(--primary);
  outline-width: 1px;
}

.main {
  composes: field from '~styles/fields.css';
  position: relative;
}

.selectInner {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.select[aria-disabled="true"], .select[aria-busy="true"] {
  opacity: 0.7;
}

.select svg {
  fill: var(--grey-4);
  stroke: var(--grey-4);
}

.select[aria-expanded="true"] i {
  transform: rotate(180deg);
}

.selectExpandContainer {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 21px;
  width: 21px;
}

.activeOption {
  height: 100%;
  overflow: hidden;
  font-weight: var(--weight-semi);
  text-overflow: ellipsis;
  white-space: nowrap;
}

.select[aria-expanded="true"] .activeOption, .select[aria-expanded="true"]:focus {
  color: var(--sky-blue);
}

/* themes */
.themeDefault {
  composes: baseSelect;
  border: 1px solid var(--temp-grey-5);
  background-color: white;
}

.themeDefault:focus {
  border-color: var(--primary);
  outline: 0;
}

.themeAlt {
  composes: baseSelect;
  padding: 5px 0 5px 5px;
  border: none;
  border-radius: var(--radius-normal);
  background-color: transparent;
  font-size: var(--size-small);
  outline: none;
}

.themeAlt.select[aria-expanded="true"], .themeAlt:focus {
  background-color: var(--temp-grey-5);
}
