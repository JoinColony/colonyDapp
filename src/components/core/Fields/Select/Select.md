
### Basic Select Input

```jsx
const { Formik } = require('formik');
const options = [
  { label: 'Option 1', value: 1 },
  { label: 'Option 2', value: 2 },
  { label: 'Option three', value: 'three' },
];
<Formik
  onSubmit={(values) => console.log(values)}
  render={({ handleSubmit }) => (
    <form onSubmit={handleSubmit}>
      <Select 
        label="I'm a Select" 
        options={options} 
        placeholder="Select an option"
        name="basicSelect"
      />
      <Select
        appearance={{ alignOptions: 'left', theme: 'alt' }}
        label="I'm an alt Select" 
        options={options} 
        placeholder="Select an alt option"
        name="selectAltTheme"
      />
      <Button type="submit">Press Me</Button>
    </form>
  )}
/>
```

### Unconnected select input

```jsx
const options = [
  { label: 'Option 1', value: 1 },
  { label: 'Option 2', value: 2 },
  { label: 'Option three', value: 'three' },
];
initialState = { $value: ''};
<div style={{width: '120px'}}>
  <Select
    connect={false}
    // with connect={false}, `$value` and `setValue` are required
    $value={state.$value}
    setValue={val => setState({ $value: val })}
    appearance={{ alignOptions: 'right', theme: 'alt' }}
    elementOnly={true}
    label="I'm an unconnected Select" 
    options={options} 
    placeholder="Select"
    name="selectUnconnected"
  />
  <Button 
    onClick={() => {
      console.log(`Currenly chosen value: ${state.$value}`);
    }}
  >
    Press Me
  </Button>
</div>
```