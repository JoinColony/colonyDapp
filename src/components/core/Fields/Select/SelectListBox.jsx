/* @flow */
import type { MessageDescriptor, MessageValues } from 'react-intl';

import React from 'react';

import { getMainClasses } from '~utils/css';

import styles from './SelectListBox.css';

import SelectOption from '../SelectOption';

import type { Appearance } from './types';

const displayName = 'SelectListBox';

type Props = {|
  appearance: Appearance,
  checkedOption: number,
  listboxId: string,
  options: Array<{
    label: MessageDescriptor | string,
    value: string,
    labelValues?: MessageValues,
  }>,
  selectedOption: number,
  ariaLabelledby: string,
  name: string,
  onSelect: (idx: number) => void,
  onClick: () => void,
  formatIntl: (
    text: string | MessageDescriptor,
    textValues?: MessageValues,
  ) => string,
|};

const getOptionId = (name, idx) =>
  idx >= 0 ? `${name}-listbox-entry-${idx}` : '';

const SelectListBox = ({
  appearance,
  listboxId,
  options,
  selectedOption,
  checkedOption,
  onSelect,
  onClick,
  formatIntl,
  ariaLabelledby,
  name,
}: Props) => {
  const activeDescendantOption = options.find(
    (_, idx) => selectedOption === idx,
  );
  const activeDescendantIdx = activeDescendantOption
    ? options.findIndex(option => option.value === activeDescendantOption.value)
    : -1;
  return (
    <ul // eslint-disable-line jsx-a11y/aria-activedescendant-has-tabindex
      tabIndex={-1}
      className={getMainClasses(appearance, styles)}
      role="listbox"
      aria-activedescendant={getOptionId(name, activeDescendantIdx)}
      id={listboxId}
      aria-labelledby={ariaLabelledby}
    >
      {options.map((option, idx) => (
        <SelectOption
          id={getOptionId(name, idx)}
          idx={idx}
          key={getOptionId(name, idx)}
          selected={selectedOption === idx}
          checked={checkedOption === idx}
          option={option}
          onSelect={onSelect}
          onClick={onClick}
          formatIntl={formatIntl}
        />
      ))}
    </ul>
  );
};

SelectListBox.displayName = displayName;

export default SelectListBox;
