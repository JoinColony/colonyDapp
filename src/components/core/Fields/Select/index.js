/* @flow */
export { default } from './Select.jsx';
export { default as SelectListBox } from './SelectListBox.jsx';
