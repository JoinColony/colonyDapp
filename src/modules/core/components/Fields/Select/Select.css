.baseSelect {
  padding: 5px 10px;
  width: auto;
  width: 100%;
  position: relative;
  border: 0;
  font-size: var(--size-normal);
  font-style: normal;
  color: var(--text);
  transition: none;
  cursor: pointer;
  outline-color: var(--primary);
  outline-width: 1px;
}

.main {
  composes: field from '~styles/fields.css';
}

.selectInner {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.inputWrapper {
  position: relative;
}

.select[aria-disabled="true"], .select[aria-busy="true"] {
  opacity: 0.7;
}

.select svg {
  fill: var(--grey-4);
  stroke: var(--grey-4);
}

.select[aria-expanded="true"] i {
  transform: rotate(180deg);
}

.selectExpandContainer {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 21px;
  width: 21px;
}

.activeOption {
  height: 100%;
  overflow: hidden;
  font-weight: var(--weight-bold);
  text-overflow: ellipsis;
  white-space: nowrap;
}

.select[aria-expanded="true"] .activeOption, .select[aria-expanded="true"]:focus {
  color: var(--sky-blue);
}

/* themes */
.themeDefault {
  composes: baseSelect;
  border: 1px solid var(--temp-grey-5);
  background-color: white;
}

.themeDefault:focus {
  border-color: var(--primary);
  outline: 0;
}

.themeAlt {
  composes: baseSelect;
  padding: 5px 0 5px 5px;
  border: none;
  border-radius: var(--radius-normal);
  background-color: transparent;
  font-size: var(--size-small);
  outline: none;
}

.themeAlt.select[aria-expanded="true"], .themeAlt:focus {
  background-color: var(--temp-grey-5);
}

.themeGrey {
  composes: baseSelect;
  padding: 10.5px 5px 10.5px 15px;
  border: 1px solid rgb(200, 214, 245);
  border-radius: var(--radius-tiny);
  background-color: rgb(249, 250, 250);
  color: rgb(60, 68, 77);
  box-shadow: inset 0px 2px 4px rgba(0, 0, 0, 0.07);
}

.themeGrey .activeOption {
  font-weight: var(--weight-normal);
}

.widthContent {
  width: min-content;
}

.widthFluid {
  width: 100%;
}

.widthStrict {
  width: 100px;
}
