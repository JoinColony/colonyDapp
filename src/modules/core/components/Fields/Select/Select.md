
### Basic Select Input

```jsx
import { Formik } from 'formik';
import Button from '../../Button';

const options = [
  { label: 'Option 1', value: 1 },
  { label: 'Option 2', value: 2 },
  { label: 'Option three', value: 'three' },
];
<Formik
  onSubmit={(values) => console.log(values)}
  render={({ handleSubmit }) => (
    <form onSubmit={handleSubmit}>
      <Select
        label="I'm a Select"
        options={options}
        placeholder="Select an option"
        name="basicSelect"
      />
      <Select
        appearance={{ alignOptions: 'left', theme: 'alt', width: 'fluid' }}
        label="I'm an alt Select"
        options={options}
        placeholder="Select an alt option"
        name="selectAltTheme"
      />
      <Button type="submit">Press Me</Button>
    </form>
  )}
/>
```

### Unconnected select input

```jsx
import Button from '../../Button';

const options = [
  { label: 'Option 1', value: 1 },
  { label: 'Option 2', value: 2 },
  { label: 'Option three', value: 'three' },
];
initialState = { $value: ''};
<div style={{width: '120px'}}>
  <Select
    connect={false}
    // with connect={false}, `$value` and `setValue` are required
    $value={state.$value}
    setValue={val => setState({ $value: val })}
    appearance={{ alignOptions: 'right', theme: 'alt', width: 'strict' }}
    elementOnly={true}
    label="I'm an unconnected Select"
    options={options}
    placeholder="Select"
    name="selectUnconnected"
  />
  <Button
    onClick={() => {
      console.log(`Currenly chosen value: ${state.$value}`);
    }}
  >
    Press Me
  </Button>
</div>
```
