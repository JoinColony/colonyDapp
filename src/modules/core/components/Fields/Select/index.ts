export { default } from './Select';
export { default as SelectListBox } from './SelectListBox';
export { Appearance, SelectOption } from './types';
export { default as SelectHorizontal } from './SelectHorizontal';
