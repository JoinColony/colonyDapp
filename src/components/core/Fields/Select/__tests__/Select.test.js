/* eslint-env jest */

import React from 'react';
import toJson from 'enzyme-to-json';

import { shallowWithIntl } from 'testutils';

import Select from '../Select.jsx';

const DEFAULT_PROPS = {
  label: 'Select me',
  name: 'foo',
  options: [
    {
      value: 'foo_value',
      label: 'Foo',
    },
    {
      value: 'bar_value',
      // Yes, it works with MessageDescriptors as well!
      label: { id: 'Select.Bar', defaultMessage: 'Bar' },
    },
  ],
  connect: false,
};

describe('Select component', () => {
  test('Renders initial component', () => {
    const wrapper = shallowWithIntl(<Select {...DEFAULT_PROPS} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
