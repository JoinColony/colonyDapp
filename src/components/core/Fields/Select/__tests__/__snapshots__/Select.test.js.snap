// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component Renders initial component 1`] = `
<Component
  connect={false}
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  label="Select me"
  name="foo"
  options={
    Array [
      Object {
        "label": "Foo",
        "value": "foo_value",
      },
      Object {
        "label": Object {
          "defaultMessage": "Bar",
          "id": "Select.Bar",
        },
        "value": "bar_value",
      },
    ]
  }
/>
`;
